backend/checkmach.ml

```diff
@@ -0,0 +1,462 @@
+    match (i.desc : Mach.instruction_desc) with
+    | Iend -> raise_after
+    | Iop op ->
+      let raise_after = check_instr_exn t i.next raise_after in
+      if not raise_after then check_operation t op i.dbg;
+      raise_after
+    | Iifthenelse (_c, ifso, ifnot) ->
+      let raise_after = check_instr_exn t i.next raise_after in
+      let raise_after_ifso = check_instr_exn t ifso raise_after in
+      let raise_after_ifnot = check_instr_exn t ifnot raise_after in
+      raise_after_ifso && raise_after_ifnot
+    | Iswitch (_index, cases) ->
+      let raise_after = check_instr_exn t i.next raise_after in
+      Array.for_all (fun case -> check_instr_exn t case raise_after) cases
+    | Icatch (_rec, _ts, handlers, body) ->
+      (* conservative *)
+      let raise_after = check_instr_exn t i.next raise_after in
+      let _ = check_instr_exn t body raise_after in
+      List.iter
+        (fun (_n, _ts, handler) -> check_instr_exn t handler false |> ignore)
+        handlers;
+      false
+    | Itrywith (body, _kind, (_ts, handler)) ->
+      (* conservative *)
+      let _ = check_instr_exn t i.next raise_after in
+      let _ = check_instr_exn t body false in
+      let _ = check_instr_exn t handler false in
+      false
+    | Iraise _ -> false
+    | Ireturn _ -> false
+    | Iexit _ -> false
+
+  let debug t expected =
+    match Unit_info.get_value unit_info t.fun_name with
+    | None -> assert false
+    | Some v ->
+      (* just checking for consistency with the recorded value *)
+      if not (v = expected)
+      then
+        Misc.fatal_errorf
+          "Checkmach %s: Wrong result for %s, expected %s, found %s" S.name
+          t.fun_name (Value.to_string expected) (Value.to_string v)
+
+  let fundecl ppf (f : Mach.fundecl) =
+    if S.enabled ()
+    then
+      Profile.record_call ~accumulate:true ("check " ^ S.name) (fun () ->
+          let fun_name = f.fun_name in
+          let t = { ppf; fun_name; unresolved_dependencies = false } in
+          Unit_info.in_current_unit unit_info fun_name;
+          (try
+             let _ = check_instr_exn t f.fun_body false in
+             if (not t.unresolved_dependencies)
+                && not (Unit_info.is_fail unit_info t.fun_name)
+             then (
+               report t ~msg:"passed" ~desc:"" f.fun_dbg;
+               Unit_info.add_value t.ppf unit_info fun_name Pass)
+           with Bail -> debug t Fail);
+          if List.mem S.annotation f.fun_codegen_options
+          then Unit_info.annotated unit_info fun_name)
+end
+
+module Spec_alloc : Spec = struct
+  let name = "alloc"
+
+  let enabled () = !Flambda_backend_flags.alloc_check
+
+  let check_callee s (checks : Cmx_format.checks) =
+    StringSet.mem s checks.ui_noalloc_functions
+
+  let add_callee s (checks : Cmx_format.checks) =
+    if not (check_callee s checks)
+    then
+      checks.ui_noalloc_functions <- StringSet.add s checks.ui_noalloc_functions
+
+  (** conservative *)
```

Comment:
I am not sure it is; do the "checkbound"
(arm64) instructions allocate? Even if
they don't, the check can be silently
broken by the addition of a new operation.

Reply:
My reasoning was if an operation allocates, it must be marked as "can_raise".

Reply:
I don't know why. My reasoning was based on if Ialloc can raise than anything else that allocations should also be marked as can_raise.

Reply:
Is it something that could somehow be affected
by polling points? (I am out of my depth.)

Reply:
Those will also need to be marked as raising, until we merge the fix for asynchronous exceptions.  At that point neither allocation points nor polling points will raise.

Reply:
Discussed off-line: we should just add a
dedicated function for allocation, even if
for now its semantics is aligned with the
one about raises.